# Simulere sannsynligheter

**Spesielt relevant kompetansemål**:

- Simulere utfall i, utforske og tolke ulike statistiske fordelinger, og gi eksempler på reelle anvendelser av disse fordelingene.

Å simulere utfall i ulike situasjoner er en effektiv måte å utforske sannsynlighetsfordelinger på, og å regne ut sannsynligheter, forvetningsverdier, varians og standardavvik i mer kompliserte situasjoner der det ikke er like lett å regne ut disse verdiene for hånd - hvis mulig i det hele tatt.

Et viktig bibliotek her er `numpy.random`, som lar oss trekke tilfeldig tall fra ulike sannsynlighetsfordelinger. 

## Hva vil det si å simulere utfall?

Å simulere utfall vil si å gjennomføre et eksperiment eller forsøk mange ganger, og telle opp hvor mange ganger de ulike utfallene inntreffer.
Som en tilnærming til sannsynligheten til et utfall, tar vi forholdet mellom antall ganger utfallet inntreffer og antall ganger vi gjennomfører eksperimentet.

Mer formelt, la $X$ være en *stokastisk* variabel (tilfeldig variabel), la $n_A$ være antall ganger et utfall $A$ skjedde, og $n$ være antall ganger vi gjennomførte eksperimentet. Da kan vi tilnærme sannsynligheten for at utfall $A$ skjer med

$$
P(X = A) \approx \frac{n_A}{n}.
$$ (eq:tilnarmet_sannsynlighet)

Da kan vi garantere at når $n$ blir veldig stor, så vil tilnærmingen bli veldig god.

```{prf:algorithm} Tilnærme sannsynligheter for utfall
:label: algo-tilnarmet-sannsynlighet
__Input__:
- $n$: Antall ganger vi gjennomfører eksperimentet
- $P(X=x)$ (diskret) eller $f(x)$ (kontinuerlig): Sannsynlighetsfordeling
- $A$: Et utfall

__Output__:
- $P(X = A)$: Tilnærmet sannsynlighet for utfall $A$

__Metode__:

- Sett $n_A = 0$
- For $i$ fra $1, 2, 3, \ldots, n$:
    - Trekk et tilfeldig tall $x$ fra sannsynlighetsfordelingen
    - Hvis $x = A$:
        - $n_A = n_A + 1$
- Returner $P(X = A) \approx \frac{n_A}{n}$
```

## Oversikt over sannsynlighetsfordelinger i `numpy.random`

Vi skal se på noen av de vanligste sannsynlighetsfordelingene, og hvordan vi kan simulere utfall fra dem.


| Sannsynlighetsfordeling | Matematisk funksjon | `numpy.random`-funksjon | Parametre | Forklaring | 
|-------------------------|---------------------|-------------------------|-----------|------------|
| Binomisk                | $B(n, p)$           | `binomial`              | `n`, `p`  | `n` angir antall forsøk og `p` angir sannsynligheten for å oppnå suksess |
| Geometrisk              | $G(p)$              | `geometric`             | `p`       | `p` angir sannsynligheten for å oppnå suksess. Er det samme som binomisk fordeling med `n = 1` |
| Normal                  | $N(\mu, \sigma) = \frac{1}{\sqrt{2\pi}\sigma} e^{-\frac{1}{2}\left((x-\mu)/\sigma\right)^2} $  | `normal`                | `loc`, `scale` | `loc` angir forventningsverdien og `scale` angir standardavviket |
| Eksponential            | $E(\lambda) = \lambda e^{-\lambda X}$        | `exponential`           | `scale`   | `scale` angir gjennomsnittlig tid mellom suksesser |
| Uniform (heltall) | $U(a, b)$ | `randint` | `low`, `high` | `low` angir minimum og `high` angir maksimum. Henter ut et heltall på intervallet `[low, high)`.  |
| Uniform (flyttall) | $U(a, b)$ | `uniform` | `low`, `high` | `low` angir minimum og `high` angir maksimum. Henter ut et flyttall på intervallet `[low, high)`.  |


---

## Eksempler på å regne ut sannsynlighet med simulering

### Eksempel 1: Terningkast med én terning

Se for deg at vi ønsker å finne sannsynligheten for å få en sekser når vi kaster en terning. Vi kan simulere dette ved å gjennomføre mange terningkast, og telle opp hvor mange ganger vi får en sekser. Så bruker vi likning {eq}`eq:tilnarmet_sannsynlighet` til å regne ut en tilnærmet verdi for terningkastet. 

I et terningkast er det 6 mulige utfall, og alle er like sannsynlige. Dermed er sannsynligheten for å få en sekser $1/6$. Lar vi $X$ være antall øyne på en terning i et tilfeldig terningkast, så er $X$ en stokastisk variabel med sannsynlighetsfordeling

|$x$|$1$|$2$|$3$|$4$|$5$|$6$|
|---|-|-|-|-|-|-|
|$P(X = x)$|$1/6$|$1/6$|$1/6$|$1/6$|$1/6$|$1/6$|

Vi kan simulere et terningkast ved å trekke ut et tilfeldig tall fra en uniform heltallsfordeling med `low=0` og `high=7`. Da kan vi få tallene 0, 1, 2, 3, 4, 5 eller 6, og vi kan tolke disse som antall øyne på terningen. Vi kan simulere et terningkast med `np.random.randint(0, 7)`, som følger:

In [1]:
import numpy as np 

n_forsøk = 100_000 # 10 000 terningkast
n_seksere = 0 # Antall seksere
for _ in range(n_forsøk):
    kast = np.random.randint(1, 7) # Kaster en terning
    if kast == 6: # Hvis vi får en sekser
        n_seksere += 1 # Legg til enda en sekser

sannsynlighet_sekser = n_seksere / n_forsøk

print(
    f"""
    Sannsynlighet for å få en sekser: {sannsynlighet_sekser:.3f}. 
    Det eksakte svaret er {1/6 :.3f} til tre desimaler.
    """
)


    Sannsynlighet for å få en sekser: 0.168. 
    Det eksakte svaret er 0.167 til tre desimaler.
    


Med andre ord er $P(X = 6) \approx 0.167$.

Siden vi trekker tilfeldige tall, og vi har begrenset oss til 100 000 forsøk, vil vi få noe varierte resultater fra gang til gang når vi kjører programmet. Med store verdier for `n_forsøk` vil resultatet bli nærmere den eksakte verdien, og det vil være lite variasjon fra gang til gang.

### Eksempel 2: Terningkast med to terninger

Vi kan også simulere et terningkast med to terninger. Da er det 36 mulige utfall, og alle er like sannsynlige. Men summen av øynene har ulike sannsynligheter. Lar vi $X$ være antall øyne vi får på to terninger i et tilfeldig terningkast, så er $X$ en stokastisk variabel med sannsynlighetsfordeling

|$x$|$2$|$3$|$4$|$5$|$6$|$7$|$8$|$9$|$10$|$11$|$12$|
|---|-|-|-|-|-|-|-|-|-|-|-|
|$P(X = x)$|$1/36$|$2/36$|$3/36$|$4/36$|$5/36$|$6/36$|$5/36$|$4/36$|$3/36$|$2/36$|$1/36$|


Tenk deg at vi ønsker å finne sannsynligheten for å få 7 øyne. Vi kan simulere dette ved å gjennomføre mange terningkast, og telle opp hvor mange ganger vi får 7 øyne. Så bruker vi likning {eq}`eq:tilnarmet_sannsynlighet` til å regne ut en tilnærmet verdi for terningkastet.

Igjen bruker vi `numpy.random.randint(1, 7)` for å kaste en terning, men nå må vi kaste to ganger per forsøk. En Pythonkode som løser dette er:

In [2]:
import numpy as np

n_forsøk = 100_000 # 100 000 terningkast med to terninger
n_suksess = 0 # Antall ganger vi får 7 som summen av øyne på terningene
for _ in range(n_forsøk):
    kast = np.random.randint(1, 7, size=2) # Kaster to terninger, derfor `size=2`
    sum_av_terninger = np.sum(kast) # Summerer terningene
    if sum_av_terninger == 7:
        n_suksess += 1
    
sannsynlighet_7 = n_suksess / n_forsøk

print(
    f"""
    Sannsynlighet for å få 7 som summen av øyne på to terninger: {sannsynlighet_7:.3f}. 
    Det eksakte svaret er {6/36 :.3f} til tre desimaler.
    """
)


    Sannsynlighet for å få 7 som summen av øyne på to terninger: 0.166. 
    Det eksakte svaret er 0.167 til tre desimaler.
    


så vi kan konkludere at $P(X = 7) \approx 0.167$.

### Eksempel 3: Normalfordelte høyder i en populasjon

Tenk deg at høydene $X$ til jenter i en populasjon er normalfordelte med forventningsverdi $\mu = 175 \ \text{cm}$ og standardavvik $\sigma = 5 \ \text{cm}$. Vi kan simulere dette ved å trekke ut tilfeldige tall fra en normalfordeling med `loc=175` og `scale=5` som parametere i `numpy.random.normal(loc=175, scale=5)`. Da vil vi få tall som er normalfordelt med forventningsverdi 175 og standardavvik 5. 

Tenk deg at vi ønsker å regne ut sannsynligheten for at en høyden til en tilfeldig jente er større enn $180 \ \text{cm}$. Dette kan vi gjøre ved å trekke ut høyder fra normalfordelingen og telle opp antaller ganger $X \geq 180 \ \text{cm}$.

En pythonkode som simulerer dette kan se slik ut:

In [4]:
import numpy as np

n_forsøk = 100_000 # Måler 100 000 høyder 
n_suksess = 0 # Antall ganger vi får en høyde er større enn 180 cm

for _ in range(n_forsøk):
    # Trekker en høyde fra normalfordeling med gjennomsnitt 175 cm og standardavvik 5 cm
    høyde = np.random.normal(loc=175, scale=5) 

    # Hvis høyden er større enn eller lik 180 cm, så legger vi til en suksess
    if høyde >= 180: 
        n_suksess += 1

sannsynlighet_høyde_over_180 = n_suksess / n_forsøk

print(f"Sannsynlighet for å få en høyde over 180 cm: {sannsynlighet_høyde_over_180:.3f}.")

Sannsynlighet for å få en høyde over 180 cm: 0.159.


som vi tolker som at $P(X \geq 180 \ \text{cm}) \approx 0.16$. 

### Eksempel 4: Normalfordelte høyder i en populasjon (igjen)

La oss fortsette med eksempel 3, men vi ønsker nå i stedet å regne ut sannsynligheten for at en høyde er mellom 170 cm og 180 cm. Fra teorien om normalfordelinger vet vi at dette er ca. 68 % siden det er sannsynligheten for å ligge ett standardavvik over eller under forventningsverdien. 

Vi kan simulere dette ved å trekke ut høyder fra normalfordelingen og telle opp antaller ganger $170 \ \text{cm} \leq X \leq 180 \ \text{cm}$:

In [16]:
import numpy as np

n_forsøk = 100_000 # Måler 100 000 høyder
n_suksess = 0 # Antall ganger vi får en høyde er større enn 180 cm

for _ in range(n_forsøk):
    # Trekker en høyde fra normalfordeling med gjennomsnitt 175 cm og standardavvik 5 cm
    høyde = np.random.normal(loc=175, scale=5) 

    # Hvis høyden ligger mellom 170 og 180 cm, så legger vi til en suksess
    if 170 <= høyde <= 180: 
        n_suksess += 1

sannsynlighet_høyde_mellom_170_og_180 = n_suksess / n_forsøk

print(f"P(170 cm <= x <= 180 cm) = {sannsynlighet_høyde_mellom_170_og_180:.3f}.")

P(170 cm <= x <= 180 cm) = 0.682.


som vi tolker som at $P(170 \ \text{cm} \leq X \leq 180 \ \text{cm}) \approx 0.68$.

### Eksempel 5: Normalfordelte høyder fra to populasjoner

Tenk deg at det på en skole går 200 jenter og 300 gutter. Jentene er normalfordelt med forventningsverdi $\mu = 168 \ \text{cm}$ og standardavvik $\sigma = 5 \ \text{cm}$, mens guttene er normalfordelt med forventningsverdi $\mu = 180 \ \text{cm}$ og standardavvik $\sigma = 6 \ \text{cm}$. 

Tenk deg at vi ønsker å regne ut sannsynligheten for at en tilfeldig trukket person har en høyde som er lavere enn 160 cm. Vi lar $X$ være høyden til en tilfeldig utvalgt elev fra skolen. Vi ønsker derfor å regne ut $P(X \leq 160 \ \text{cm})$.

Det som er vrient her, er at det ikke er så lett å sette opp en sannsynlighetsfordeling fordi vi har en normalfordeling for den ene delen av populasjon (guttene), og en annen normalfordeling for den andre delen (jentene). 

Vi skal bryte ned problemet i to biter. 

1. Hvordan velge om vi har trukket ut en jente eller en gutt.
2. Deretter trekke et tilfeldig høyde avhengig om det var en gutt eller jente som ble trukket ut.



#### Hvordan velge ut en gutt eller en jente fra elevene på skolen

La $N_\text{jenter} = 200$ og $N_\text{gutter} = 300$. Det totale antallet elever er $N_\text{elever} = N_\text{jenter} + N_\text{gutter} = 500$. Sannsynligheten for at vi en tilfeldig utvalgt elev er en jente er da

$$
p_\text{jente} = \frac{N_\text{jenter}}{N_\text{elever}} = \frac{200}{500} = \frac{2}{5} = 0.4. 
$$

Vi kan betrakte dette som en binomisk fordeling med $n = 1$ og $p = p_\text{jente}$.

La $X$ være antall jenter vi trekker ut. Da er $X$ binomisk fordelt med $n = 1$ og $p = p_\text{jente}$. Da vil $X = 1$ dersom vi trekker ut en jente, og $X = 0$ dersom vi ikke gjør det (og dermed trekker en gutt i stedet). 
Vi kan simulere dette ved å bruke `numpy.random.binomial(n=1, p=p_jente)`.

La oss sjekke at dette gir mening:

In [15]:
n_forsøk = 100_000 # Antall elever vi trekker ut
n_suksess = 0 # Antall ganger vi trekker ut en jente
n_jenter = 200
n_gutter = 300
n_elever = n_jenter + n_gutter
p_jente = n_jenter / n_elever # Sannsynlighet for å trekke ut en jente

for _ in range(n_forsøk):
    # Trekker en elev, der 1 er jente og 0 er gutt
    elev = np.random.binomial(n=1, p=p_jente) 

    # Hvis vi trekker ut en jente
    if elev == 1: 
        n_suksess += 1 # Legg til en suksess
    
sannsynlighet_jente = n_suksess / n_forsøk

print(
    f"""
    Sannsynlighet for å trekke ut en jente: {sannsynlighet_jente:.3f}. 
    Det eksakte svaret er {n_jenter / n_elever :.3f} til tre desimaler.
    """
)



    Sannsynlighet for å trekke ut en jente: 0.401. 
    Det eksakte svaret er 0.400 til tre desimaler.
    


Så vi får bekreftet at $P(X = \text{jente}) = p_\text{jente} = 0.4$.

#### Hvordan trekke en høyde avhengig av om det er en gutt eller jente

Nå har vi funnet ut hvordan vi kan velge om vi har trukket ut en jente eller en gutt. Nå skal vi finne ut hvordan vi kan trekke en høyde avhengig av om det er en jente eller gutt.

1. Hvis vi har en jente, trekker vi en høyde fra en normalfordeling med forventningsverdi $\mu_\text{jente} = 168 \ \text{cm}$ og standardavvik $\sigma_\text{jente} = 6 \ \text{cm}$.
2. Hvis det er en gutt trekker vi en høyde fra en normalfordeling med forventningsverdi $\mu_\text{gutt} = 180 \ \text{cm}$ og standardavvik $\sigma_\text{gutt} = 8 \ \text{cm}$.

#### Simulering

Vi er nå klare til å sette det hele sammen. Algoritmen vi skal bruke går som følger:

1. Velg om vi har trukket ut en jente eller en gutt.
    - Hvis vi har trukket ut en jente, trekker vi en høyde fra en normalfordeling med forventningsverdi $\mu_\text{jente} = 168 \ \text{cm}$ og standardavvik $\sigma_\text{jente} = 6 \ \text{cm}$.
    - Hvis det er en gutt trekker vi en høyde fra en normalfordeling med forventningsverdi $\mu_\text{gutt} = 180 \ \text{cm}$ og standardavvik $\sigma_\text{gutt} = 8 \ \text{cm}$.
2. Dersom høyden er under 160 cm, legger vi til 1 på $n_\text{suksess}$. 
3. Gjenta 1 og 2 $n_\text{forsøk}$ ganger.

Til slutt regner vi ut sannsynligheten med likning {eq}`eq:tilnarmet_sannsynlighet` som 

$$
P(X \leq 160 \ \text{cm}) \approx \frac{n_\text{suksess}}{n_\text{forsøk}}.
$$

Pythonkoden kan da se slik ut:

In [17]:
import numpy as np

n_forsøk = 1000_000 # Antall elever vi trekker ut
n_suksess = 0 # Antall ganger vi får en høyde <= 160 cm.
n_jenter = 200
n_gutter = 300
n_elever = n_jenter + n_gutter
p_jente = n_jenter / n_elever # Sannsynlighet for å trekke ut en jente

for _ in range(n_forsøk):
    # Trekker en elev, der 1 er jente og 0 er gutt
    elev = np.random.binomial(n=1, p=p_jente) 
    if elev == 1: # Hvis vi trekker ut en jente
        # Trekker en høyde fra normalfordeling med gjennomsnitt 175 cm og standardavvik 5 cm
        høyde = np.random.normal(loc=168, scale=6) 
    else:
        # Trekker en høyde fra normalfordeling med gjennomsnitt 180 cm og standardavvik 6 cm
        høyde = np.random.normal(loc=180, scale=8) 
    
    if høyde <= 160: # Hvis høyden er mindre eller lik 160 cm, så legger vi til en suksess
        n_suksess += 1

sannsynlighet_høyde_under_160 = n_suksess / n_forsøk

print(f"{sannsynlighet_høyde_under_160 = }")


sannsynlighet_høyde_under_160 = 0.04057


Så sannsynligheten for at en tilfeldig utvalgt elev har en høyde som er lavere enn 160cm er 

$$
P(X \leq 160 \ \text{cm}) \approx 0.04 = 4 \%.
$$

---

## Øvingsoppgaver

### Oppgave 1

Du skal kaste to terninger. Skriv en Pythonkode for å regne ut sannsynligheten for at man får 9 eller flere øyne.

Gjør 100 000 forsøk i simuleringen din.

*Du kan ta utgangspunkt i kodeskallet under. Du må fylle inn der det står `NotImplemented`.*

In [ ]:
import numpy as np

n_forsøk = NotImplemented # Sett antall terningkast
n_suksess = 0

for _ in range(n_forsøk):
    kast = NotImplemented
    sum_av_terninger = NotImplemented
    if NotImplemented: # Sett betingelse for når du skal øke antall suksesser
        n_suksess += 1

p = NotImplemented # Regn ut sannsynligheten 

print(f"{p = }")

`````{dropdown} Løsningsforslag 
````{tab} For-løkke
```python
import numpy as np

n_forsøk = 100_000
n_suksess = 0

for _ in range(n_forsøk):
    kast = np.random.randint(1, 7, size=2)
    sum_av_terninger = np.sum(kast)
    if sum_av_terninger >= 9:
        n_suksess += 1

p = n_suksess / n_forsøk

print(f"{p = }")
```
````
````{tab} Vektorisert med Numpy
```python
import numpy as np

n_forsøk = 100_000
kast = np.random.randint(1, 7, size=(n_forsøk, 2)) # Kaster to terninger, derfor `size=(n_forsøk, 2)`
sum_av_terninger = np.sum(kast, axis=1)
n_suksess = np.sum(sum_av_terninger >= 9) 

p = n_suksess / n_forsøk

print(f"{p = }")
```
````

som gir utskriften

```console
p = 0.2757
```

Altså er $P(X \geq 9) \approx 0.2757$.

*Merk at siden vi trekker tilfeldige tall her, vil svaret man får variere litt fra gang til gang.

`````

### Oppgave 2

Levetiden til en tilfeldig valgt lyspære er normalfordelt med forventning på 400 timer og et standardavvik på 10 timer. 
Skriv en Pythonkode for å regne ut sannsynligheten for at en tilfeldig valgt lyspære har en levetid på mer enn 420 timer.


*Du kan ta utgangspunkt i kodeskallet under. Du må fylle inn der det står `NotImplemented`.*

In [ ]:
import numpy as np

n_forsøk = NotImplemented # Antall pærer vi sjekker levetiden til
n_suksess = NotImplemented # Antall ganger vi får en levetid på minst 420 timer.

for _ in range(n_forsøk):
    levetid = NotImplemented
    if NotImplemented: # Sett betingelse for når du skal øke antall suksesser
        n_suksess = NotImplemented
    
sannsynlighet_levetid_over_420 = NotImplemented # Regn ut sannsynligheten

print(f"{sannsynlighet_levetid_over_420 = :.3f}")

In [4]:
import numpy as np

n_forsøk = 100_000 # Antall pærer vi sjekker levetiden til

# Trekker ut 100 000 levetider fra normalfordeling med gjennomsnitt 400 timer og standardavvik 10 timer
levetider = np.random.normal(loc=400, scale=10, size=n_forsøk) 

# Sjekker hvor mange av levetidene som er over 420 timer
n_suksess = np.sum(levetider > 420)

sannsynlighet_levetid_over_420 = n_suksess / n_forsøk

print(f"{sannsynlighet_levetid_over_420 = :.3f}")

sannsynlighet_levetid_over_420 = 0.023


`````{dropdown} Løsningsforslag

````{tab} For-løkke
```python
import numpy as np

n_forsøk = 100_000 # Antall pærer vi sjekker levetiden til
n_suksess = 0 # Antall ganger vi får en levetid på minst 420 timer.

for _ in range(n_forsøk):
    levetid = np.random.normal(loc=400, scale=10)
    if levetid >= 420:
        n_suksess += 1
    
sannsynlighet_levetid_over_420 = n_suksess / n_forsøk

print(f"{sannsynlighet_levetid_over_420 = :.3f}")
```
````

````{tab} Vektorisert med Numpy
```python
import numpy as np

n_forsøk = 100_000 # Antall pærer vi sjekker levetiden til

# Trekker ut 100 000 levetider fra normalfordeling med gjennomsnitt 400 timer og standardavvik 10 timer
levetider = np.random.normal(loc=400, scale=10, size=n_forsøk) 

# Sjekker hvor mange av levetidene som er over 420 timer
n_suksess = np.sum(levetider > 420)

sannsynlighet_levetid_over_420 = n_suksess / n_forsøk

print(f"{sannsynlighet_levetid_over_420 = :.3f}")
```
````

som gir utskriften

```console
sannsynlighet_levetid_over_420 = 0.022
```

som betyr at det $P(X \geq 420 \ \text{timer}) \approx 0.02 = 2 \%.$ 

`````

### Oppgave 3

En kjede med pærer er koblet i serie hengt opp rundt et juletre midt i Karl Johan. Hver lyspære har individuelt en levetid som er normalfordelt med en forventning på 900 timer og et standardavvik på 50 timer. 

Dersom én lyspære slutter å lyse, vil alle lyspærene i kjeden slutte å lyse. Oslo Kommune er interessert i å vite hvor lenge det er sannsynlig at juletreet vil lyse. De estimerer at de setter opp juletreet i midten av desember og at de tar det ned i midten av januar. Juletreet består av 2000 lyspærer.

Målet her er å finne ut av hvor sannsynlig det er at juletreet vil lyse i hele perioden det er satt opp. Vi kan anta at alle lyspærene lyser samtidig.

#### Oppgave 3a 

Skriv en Pythonkode som regner ut sannsynlighet for at én tilfeldig valgt lyspære vil fungere minst like lenge som juletreet er satt opp.

*Du kan ta utgangspunkt i kodeskallet under. Du må fylle inn der det står `NotImplemented`.*

In [ ]:
import numpy as np

n_forsøk = NotImplemented 
n_suksess = NotImplemented
minst_levetid = NotImplemented # Perioden vi ønsker at pærene minst skal vare.

for _ in range(n_forsøk):
    levetid = NotImplemented
    if NotImplemented: 
        n_suksess = NotImplemented
    
p = NotImplemented

print(f"{p = }")

`````{dropdown} Løsningsforslag

````{tab} For-løkke
```python
import numpy as np

n_forsøk = 100_000
n_suksess = 0
minst_levetid = 30 * 24 # 30 dager * 24 timer per dag 

for _ in range(n_forsøk):
    levetid = np.random.normal(loc=900, scale=50)
    if levetid >= minst_levetid: 
        n_suksess += 1
    
p = n_suksess / n_forsøk

print(f"{p = }")
```
````

````{tab} Vektorisert med Numpy
```python
import numpy as np

n_forsøk = 100_000
minst_levetid = 24 * 30 # Perioden vi ønsker at pærene minst skal vare.

# Trekker ut 100 000 levetider fra normalfordeling med gjennomsnitt 900 timer og standardavvik 50 timer
levetider = np.random.normal(loc=900, scale=50, size=n_forsøk)

# Sjekker hvor mange av levetiden som varer minst hele perioden
n_suksess = np.sum(levetider >= minst_levetid)

# Regner ut sannsynligheten 
p = n_suksess / n_forsøk

print(f"{p = }")
```
````

som gir utskriften

```console
p = 0.99989
```

så vi kan være rimelig sikre på at én enkelt pære vil vare hele perioden.

`````

#### Oppgave 3b

Bruk Python til å regne ut sannsynligheten for at alle 2000 lyspærer varer minst hele perioden treet er satt opp.

*Du kan ta utgangspunkt i kodeskallet under. Du må fylle inn der det står `NotImplemented`.*



````{dropdown} Kodehint

Du kan trekke 2000 tilfeldige tall fra en normalfordeling ved å bruke 
```python
levetider = np.random.normal(loc=900, scale=50, size=2000)
``` 
Hvis du ønsker å sjekke om alle sammen oppfyller en betingelse samtidig, kan du bruke 

```python
if np.all(betingelse):
    # Kodeblokk her.
```

````


`````{dropdown} Løsningsforslag

*Merk at koden som kjører "ren Python" er veldig treg. De to vektoriserte versjonene er omtrent like raske.*

````{tab} Ren Python
```python
import numpy as np 

antall_pærer = 2000
n_forsøk = 100_000
n_fail = 0
minst_levetid = 30 * 24 # 30 dager * 24 timer per dag

for _ in range(n_forsøk):
    for j in range(antall_pærer):
        levetid = np.random.normal(loc=900, scale=50)
        if levetid < minst_levetid:
            n_fail += 1
            break

n_suksess = n_forsøk - n_fail
    
p = n_suksess / n_forsøk

print(f"{p = }")
```

````

````{tab} Delvis vektorisert med Numpy

```python
import numpy as np 

antall_pærer = 2000
n_forsøk = 100_000
n_suksess = 0
minst_levetid = 30 * 24 # 30 dager * 24 timer per dag

for _ in range(n_forsøk):
    levetid = np.random.normal(loc=900, scale=50, size=antall_pærer)
    if np.all(levetid >= minst_levetid): 
        n_suksess += 1
    
p = n_suksess / n_forsøk

print(f"{p = }")
```
````

````{tab} Vektorisert med Numpy
```python
import numpy as np 

antall_pærer = 2000
n_forsøk = 100_000
n_suksess = 0
minst_levetid = 30 * 24 # antall timer vi ønsker at pærene skal vare

levetider = np.random.normal(loc=900, scale=50, size=(n_forsøk, antall_pærer))

n_suksess = np.sum(np.all(levetider >= minst_levetid, axis=1))

p = n_suksess / n_forsøk

print(f"{p = }")
```
````

som gir utskriften

```console
p = 0.72904
```

som betyr at sannsynligheten for at juletreet vil lyse gjennom hele perioden er ca. 73 %.

`````